modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsConfigurationManager.java

```diff
@@ -0,0 +1,518 @@
+        Set<String> colToRemove
+    ) {
+        // Drop all columns
+        if (colToRemove.isEmpty())
+            return EMPTY_COLUMN_CFGS_ARR;
+
+        Set<StatisticsColumnConfiguration> cols = Arrays.stream(oldCfg.columns())
+            .filter(c -> !colToRemove.contains(c.name())).collect(Collectors.toSet());
+
+        // All not-hidden fields are dropped -> Drop all columns
+        if (cols.size() == 2 && cols.contains(QueryUtils.KEY_FIELD_NAME) && cols.contains(QueryUtils.VAL_FIELD_NAME))
+            return EMPTY_COLUMN_CFGS_ARR;
+
+        return cols.toArray(EMPTY_COLUMN_CFGS_ARR);
+    }
+
+    /** Drop All statistics. */
+    public void dropAll() {
+        mgmtPool.submit(() -> {
+            try {
+                final List<StatisticsTarget> targetsToRemove = new ArrayList<>();
+                distrMetaStorage.iterate(STAT_OBJ_PREFIX, (k, v) -> {
+                        StatisticsKey statKey = ((StatisticsObjectConfiguration)v).key();
+
+                        targetsToRemove.add(new StatisticsTarget(statKey.schema(), statKey.obj()));
+                    }
+                );
+
+                dropStatistics(targetsToRemove);
+            }
+            catch (IgniteCheckedException e) {
+                log.warning("Unexpected exception on check local statistic on start", e);
+            }
+        });
+    }
+
+    /** */
+    private void validate(StatisticsTarget target, GridH2Table tbl) {
+        if (tbl == null) {
+            throw new IgniteSQLException(
+                "Table doesn't exist [schema=" + target.schema() + ", table=" + target.obj() + ']');
+        }
+
+        for (String col : target.columns()) {
+            if (tbl.getColumn(col) == null) {
+                throw new IgniteSQLException(
+                    "Column doesn't exist [schema=" + target.schema() + ", table=" + target.obj() + ", column=" + col + ']');
+            }
+        }
+    }
+
+    /**
+     *
+     */
+    private void updateObjectStatisticConfiguration(StatisticsObjectConfiguration cfg) {
+        try {
+            while (true) {
+                String key = key2String(cfg.key());
+
+                StatisticsObjectConfiguration oldCfg = distrMetaStorage.read(key);
+
+                if (oldCfg != null)
+                    cfg = StatisticsObjectConfiguration.merge(oldCfg, cfg);
+
+                if (distrMetaStorage.compareAndSet(key, oldCfg, cfg))
+                    return;
+            }
+        }
+        catch (IgniteCheckedException ex) {
+            throw new IgniteSQLException("Error on get or update statistic schema", IgniteQueryErrorCode.UNKNOWN, ex);
+        }
+    }
+
+    /** */
+    private void checkLocalStatistics(StatisticsObjectConfiguration cfg, final AffinityTopologyVersion topVer) {
+        try {
+            GridH2Table tbl = schemaMgr.dataTable(cfg.key().schema(), cfg.key().obj());
+
+            if (tbl == null) {
+                // Drop tables handle by onDropTable
```

Comment:
If the node was out of the cluster while table dropped - it still get those dropTable event?

Reply:
Invalid case. The exception is thrown:
`Caused by: class org.apache.ignite.spi.IgniteSpiException: Joining node has caches with data which are not presented on cluster, it could mean that they were already destroyed, to add the node to cluster - remove directories with the caches[SMALL]
	at org.apache.ignite.spi.discovery.tcp.TcpDiscoverySpi.checkFailedError(TcpDiscoverySpi.java:2031)
	at org.apache.ignite.spi.discovery.tcp.ServerImpl.joinTopology(ServerImpl.java:1174)
	at org.apache.ignite.spi.discovery.tcp.ServerImpl.spiStart(ServerImpl.java:445)
	at org.apache.ignite.spi.discovery.tcp.TcpDiscoverySpi.spiStart(TcpDiscoverySpi.java:2133)
	at org.apache.ignite.internal.managers.GridManagerAdapter.startSpi(GridManagerAdapter.java:277)
`